Stand-alone testing: remove deprecated methods

AcriusWinter opened the pull request:
Make changes as requested by #34236 to get rid of deprecated methods

WARNING: Depends on the conversion/removal of the use of all deprecated methods from built-in and test packages.

TODO

- [x] Here **_or another PR_** .. add audit check for test_ methods to have docstrings
- [x] Add audit checks that none of the deprecated methods get added again

AcriusWinter committed:
Stand-alone testing: remove deprecated methods
* audit: replace deprecated test method checks for test callbacks (AcriusWinter)
* install_test: replace deprecated with new test method name starts (AcriusWinter)
* package_base: removed deprecated test methods (AcriusWinter)
* test/package_class: remove deprecated test methods (AcriusWinter)
* test/reporters: remove deprecated test methods (AcriusWinter)
* reporters/extract: remove deprecated test-related methods (AcriusWinter)
* py-test-callback: rename test in callbacks and output (AcriusWinter)
* reporters/extract: remove deprecated expected failure output capture (AcriusWinter)
* stand-alone test cleanup: f-string, remove deprecation warning, etc. (AcriusWinter)
* stand-alone tests: f-string fix and remove unused imports (AcriusWinter)
* package_base: finish removing the rest of deprecated run_test method (AcriusWinter)
* py-test-callback: remove stand-alone test method (AcriusWinter)
* package_base: remove now unused imports (AcriusWinter)
* Python: test_imports replaces test (tldahlgren)
* mptensor, trivial-smoke-test: replace deprecated cache_extra_test_sources method (tldahlgren)
* trivial-smoke-test: replace deprecated install_test_root method (tldahlgren)
* docs: update perl and scons package's test methods (tldahlgren)
* builder: remove deprecated test() method (tldahlgren)
* Update legion and mfem test data for stand-alone tests (tldahlgren)

tldahlgren commented on lib/spack/spack/audit.py:
```diff
@@ -416,12 +416,10 @@ def _check_build_test_callbacks(pkgs, error_cls):
-            msg = '{0} package contains "test*" method(s) in ' "build_time_test_callbacks"
-            instr = 'Remove all methods whose names start with "test" from: [{0}]'.format(
+            msg = '{0} package contains "test_*" method(s) in ' "build_time_test_callbacks"
+            instr = 'Remove all methods whose names start with "test_" from: [{0}]'.format(
```
This should be an f-string.

tldahlgren commented on lib/spack/spack/audit.py:
```diff
@@ -416,12 +416,10 @@ def _check_build_test_callbacks(pkgs, error_cls):
-        # TODO (post-34236): "test*"->"test_*" once remove deprecated methods
-        # TODO (post-34236): "test"->"test_" once remove deprecated methods
-        has_test_method = test_callbacks and any([m.startswith("test") for m in test_callbacks])
+        has_test_method = test_callbacks and any([m.startswith("test_") for m in test_callbacks])
```
🤔 I know I flagged it for change (`test` ->  `test_`) in the mentioned PR but I do wonder if it is a good idea since I can imagine ppl being confused if their `test` method doesn't get executed. Definitely something to bring up in a discussion with others.

tldahlgren replied:
Upon further reflection, perhaps it is fine to keep this change.

tldahlgren commented on lib/spack/spack/install_test.py:
```diff
@@ -468,12 +467,11 @@ def write_tested_status(self):
-    ), f"Test name must start with 'test' but {test_name} was provided"
+        "test_"
+    ), f"Test name must start with 'test_' but {test_name} was provided"
 
```
This should be removed.  No-one will have `def test_(self, ...)` defined to warn them about.

tldahlgren commented on lib/spack/spack/reporters/extract.py:
```diff
@@ -153,13 +130,9 @@ def extract_test_parts(default_name, outputs):
-                    if part:
-                        part["desc"] += f"; {msg}"
-                    continue
+                if part:
+                    part["desc"] += f"; {msg}"
+                continue
```
The goal of the original check was to determine whether message processing was done (and append the expectation to the description).  In this case, since the new API does not support expected return codes, it is not.  So the rest of the check needs to be removed.

tldahlgren commented on var/spack/repos/builtin.mock/packages/py-test-callback/package.py:
```diff
@@ -13,17 +13,15 @@ class PyTestCallback(mp.Python):
         mkdirp(prefix.bin)
 
-    # TODO (post-34236): "test" -> "test_callback" once remove "test" support
     def test(self):
```
Looks like you forgot to rename the `test` method per the note.

tldahlgren reviewed (changes requested):
This is a great start!  I see several issues with the current state of the draft that I've flagged.

tldahlgren commented on lib/spack/spack/install_test.py:
```diff
@@ -468,12 +467,11 @@ def write_tested_status(self):
+    ), f"Test name must start with 'test_' but {test_name} was provided"
 
-    if test_name == "test":
+    if test_name == "test_":
         tty.warn(
             "{}: the 'test' method is deprecated. Convert stand-alone "
             "test(s) to methods with names starting 'test_'.".format(pkg.name)
```
(Second attempt. Hopefully properly highlighted this time.)

The check (lines 474-477) should be removed.  No-one will have `def test_(self, ...)` defined to warn them about.

tldahlgren commented on var/spack/repos/builtin.mock/packages/py-test-callback/package.py:
```diff
@@ -13,17 +13,15 @@ class PyTestCallback(mp.Python):
-    # TODO (post-34236): "test" -> "test_callback" once remove "test" support
-    def test(self):
+    def test_callback(self):
         super().test()
```
The base class no longer has a `test()` method.

tldahlgren commented on lib/spack/spack/package_base.py:
```diff
@@ -1947,51 +1902,6 @@ def do_test(self, dirty=False, externals=False):
-            work_dir (str or None): path to the smoke test directory
-        """
-
         def test_title(purpose, test_name):
```
The implementation of `run_test()` needs to be removed.

tldahlgren commented on var/spack/repos/builtin.mock/packages/py-test-callback/package.py:
```diff
@@ -13,17 +13,10 @@ class PyTestCallback(mp.Python):
-    def test(self):
-        super().test()
-
-        print("PyTestCallback test")
```
This method should not have been removed but renamed to match the name in `install_time_test_callbacks` *and* the call to `super`'s `test()` removed.

tldahlgren committed:
py-test-callback: restore the test method

tldahlgren reviewed (changes requested):
See comment(s).

tldahlgren commented on lib/spack/spack/audit.py:
```diff
@@ -408,14 +408,14 @@ def _check_build_test_callbacks(pkgs, error_cls):
+                "build_time_test_callbacks"
             )
+            instr = 'Remove all methods whose names start with "test_"'
+            instr = instr + f' from: [{", ".join(test_callbacks)}]'
```
This is not the correct f-string replacement for the original string.

tldahlgren reviewed (changes requested):
See comment.

tldahlgren committed:
Audit/stand-alone testing: Fix and added stand-alone test audit checks

- fix audit failure message for build-time test callback check
- remove empty test method check during stand-alone testing
- change build-time test callback check to package_properties
- add test method docstring audit check and mock fail-test-audit-docstring pkg
- add test method implementation audit check and mock fail-test-audit-impl pkg
- add new mock packages to test_package_audits and test_test_list_all

tldahlgren committed:
audits: loosen docstring content constraints

tldahlgren committed:
Add missing test method docstrings

tldahlgren committed:
py-test-callback: resolve builder issues

tldahlgren committed:
Audits: Add checks for use of deprecated stand-alone test methods; improve output for other test-related checks

tldahlgren committed:
Fix style issues

tldahlgren committed:
test_test_list_all: add new fail-test-audit-deprecated package

tldahlgren commented:
Looks like 3 stacks have problems with `paraview` (e.g., timing out).

tldahlgren commented:
Unclear if it matters but let's try to overcome the timeouts (again).

@spackbot run pipeline

alalazo commented:
> Looks like 3 stacks have problems with paraview (e.g., timing out).

See https://github.com/spack/spack/pull/46712
